pkg/urbit/noun/retrieve.c

```diff
@@ -1029,63 +1029,46 @@ c3_w
 u3r_met(c3_y  a_y,
         u3_atom b)
 {
-  c3_assert(u3_none != b);
-  c3_assert(_(u3a_is_atom(b)));
+  c3_dessert(u3_none != b);
+  c3_dessert(_(u3a_is_atom(b)));
 
   if ( b == 0 ) {
     return 0;
   }
-  else {
-    /* gal_w: number of words besides (daz_w) in (b).
-    ** daz_w: top word in (b).
-    */
-    c3_w gal_w;
-    c3_w daz_w;
-
-    if ( _(u3a_is_cat(b)) ) {
-      gal_w = 0;
-      daz_w = b;
-    }
-    else {
-      u3a_atom* b_u = u3a_to_ptr(b);
-
-      gal_w = (b_u->len_w) - 1;
-      daz_w = b_u->buf_w[gal_w];
-    }
-
-    switch ( a_y ) {
-      case 0:
-      case 1:
-      case 2: {
-        /* col_w: number of bits in (daz_w)
-        ** bif_w: number of bits in (b)
-        */
-        c3_w bif_w, col_w;
-
-        if ( gal_w > ((UINT32_MAX - 35) >> 5) ) {
-          return u3m_bail(c3__fail);
-        }
-
-        col_w = c3_bits_word(daz_w);
-        bif_w = col_w + (gal_w << 5);
-
-        return (bif_w + ((1 << a_y) - 1)) >> a_y;
-      }
-
-      STATIC_ASSERT((UINT32_MAX > ((c3_d)u3a_maximum << 2)),
-                    "met overflow");
+  /* gal_w: number of words besides (daz_w) in (b).
+  ** daz_w: top word in (b).
+  */
+  c3_w gal_w;
+  c3_w daz_w;
 
-      case 3: return (gal_w << 2) + ((c3_bits_word(daz_w) +  7) >> 3);
+  if ( _(u3a_is_cat(b)) ) {
+    gal_w = 0;
+    daz_w = b;
+  }
+  else {
+    u3a_atom* b_u = u3a_to_ptr(b);
 
-      case 4: return (gal_w << 1) + ((c3_bits_word(daz_w) + 15) >> 4);
+    gal_w = (b_u->len_w) - 1;
+    daz_w = b_u->buf_w[gal_w];
+  }
 
-      default: {
-        c3_y gow_y = (a_y - 5);
+  /* 5 because 1<<2 bytes in c3_w, 1<<3 bits in byte.
+     aka log2(CHAR_BIT * sizeof gal_w)
+     a_y < 5 informs whether we shift return left or right
+     */
+  if (a_y < 5) {
+    c3_y max_y = (1 << a_y) - 1;
+    if (gal_w > ((UINT32_MAX - (32 + max_y)) >> (5 - a_y)))
+      return u3m_bail(c3__fail);
 
-        return ((gal_w + 1) + ((1 << gow_y) - 1)) >> gow_y;
-      }
-    }
+    return (gal_w << (5 - a_y))
+      + (c3_bits_word(daz_w)
+         + max_y
+         >> a_y);
   }
+  c3_y gow_y = (a_y - 5);
+  return ((gal_w + 1) + ((1 << gow_y) - 1)) >> gow_y;
```

Comment:
I'm not sure this shouldn't just be `gal_w + 1 >> gow_y;` A fakezod boots and runs fine with that definition. Though compared to the previous implementation, this _does_ return different results. Yes, differing results are encountered in the boot process - and rather quickly - so it's not a matter of incomplete test input.

Reply:
@joemfb Is there some hoon I could use to catastrophically crash this if the line referenced were changed to `gal_w + 1 >> gow_y;`? I'm just surprised a fakezod booted fine

Reply:
And note comments in https://github.com/urbit/urbit/pull/6077/commits/a10e50ac210a022a7a5c72cd7b6457528ed3259a . In current implementation, this _does_ match exactly

Reply:
You could turn on jet equivalence testing for `+met` (in `jets/tree.c`, change `c3y` to `c3n` for `_140_two_met_a`). Or you could boot a fake ship with one version of `u3r_met()`, then delete the snapshot replay it with the other and watch for the mug mismatches.

Reply:
Generally this kind of thing is hard (but not impossible) to detect from hoon -- you're messing with the laws of physics down here.

Reply:
Thanks! Actually, I guess I can't say it's an exact match when I also say:

""
Moreover, this change does not artificially limit the size that gal_w can
be. For instance, in the previous implementation, for a value of a_y=2, gal_w
could not exceed the following without bailing:

`(UINT32_MAX - 35) >> 5` =>
`0x07FFFFFF`

Now, gal_w cannot exceed

`((UINT32_MAX - (32 + max_y)) >> (5 - a_y))` =>
`((UINT32_MAX - 35) >> 3)` =>
`0x1FFFFFFF`
""

Maybe "an exact match" was the actual reason for the prior switch case implementation and not optimization?

Reply:
So a jet can bail when the equivalent nock wouldn't? In that case, there shouldn't be anything wrong with my change, though I'm running it through the testing you suggested.

Reply:
Ok, confirming that the following doesn't show any mug mismatches:

1) boot fakezod with current branch (`u3r_met` changes)
2) delete chk and bhk
3) recompile with old definition of `u3r_met`
4) recompute events -- no mug mismatches

And vice-versa

Though, given your comments above, I certainly _would_ expect mug mismatches if I replaced the last return with `gal_w + 1 >> gow_y;`

Reply:
Aside from generalizing and reducing instruction size, the only functional change the new `u3r_met` introduces is it won't bail as early for `a_y` values of 1 or 2.

For instance, for `a_y=2`, previous would bail when `gal_w` exceeds `0x07FFFFFF`. Now it would bail when `gal_w` exceeds `0x1FFFFFFF` (~4x the allowable length)